src/github/githubRepository.ts

```diff
@@ -4,35 +4,72 @@
  *--------------------------------------------------------------------------------------------*/
 
 import * as vscode from 'vscode';
-import * as Github from '@octokit/rest';
+import * as Octokit from '@octokit/rest';
 import Logger from '../common/logger';
 import { Remote, parseRemote } from '../common/remote';
 import { PRType, IGitHubRepository, PullRequest } from './interface';
 import { PullRequestModel } from './pullRequestModel';
-import { CredentialStore } from './credentials';
+import { CredentialStore, GitHub } from './credentials';
 import { AuthenticationError } from '../common/authentication';
+import { QueryOptions, MutationOptions, ApolloQueryResult } from 'apollo-boost';
 
 export const PULL_REQUEST_PAGE_SIZE = 20;
 
+const GRAPHQL_COMPONENT_ID = 'GraphQL';
+
 export interface PullRequestData {
 	pullRequests: PullRequestModel[];
 	hasMorePages: boolean;
 }
 
 export class GitHubRepository implements IGitHubRepository {
 	static ID = 'GitHubRepository';
-	private _octokit: Github;
+	private _hub: GitHub;
 	private _initialized: boolean;
 	private _metadata: any;
-	public get octokit(): Github {
-		if (this._octokit === undefined) {
+	public get hub(): GitHub {
+		if (this._hub === undefined) {
 			if (!this._initialized) {
 				throw new Error('Call ensure() before accessing this property.');
 			} else {
 				throw new AuthenticationError('Not authenticated.');
 			}
 		}
-		return this._octokit;
+		return this._hub;
+	}
+
+	supportsGraphQl(): boolean {
+		return !!(this.hub && this._hub.graphql);
+	}
+
+	public get octokit(): Octokit {
+		return this.hub && this._hub.octokit;
+	}
+
+	query = async <T = any>(query: QueryOptions): Promise<ApolloQueryResult<T>> => {
+		const gql = this.hub && this._hub.graphql;
+		if (!gql) {
+			Logger.debug(`Not available for query: ${query}`, GRAPHQL_COMPONENT_ID);
+			return null;
+		}
+
+		Logger.debug(`Request: ${JSON.stringify(query, null, 2)}`, GRAPHQL_COMPONENT_ID);
+		const rsp = await gql.query<T>(query);
+		Logger.debug(`Response: ${JSON.stringify(rsp, null, 2)}`, GRAPHQL_COMPONENT_ID);
+		return rsp;
+	}
+
+	mutate = async <T = any>(mutation: MutationOptions): Promise<ApolloQueryResult<T>> => {

```

Comment:
any reason that we don't declare these methods as `async mutate<T = any>(): Promise { }` ? They look foreign to all other methods we have.

Reply:
in calling code we were using the same syntax as before with ensure, so something like `const { query } = await (pullRequest as PullRequestModel).githubRepository.ensure();`. When using the method like that, the correct `this` is lost unless using the arrow function syntax or using `bind`.

I agree that these look foreign :/ we could change them to normal syntax and in calling code, just call ensure and then `(pullRequest as PullRequestModel).githubRepository.query(...)` instead